### DISCUSSION TITLE
Using two buttons to change output octave of NoteButton buttons

### DISCUSSION THREAD

**Discussion by dstoddart** [tttapa/Control-Surface#480] [Q&A]
Hello everyone, this library is amazing! This community has cut countless hours off of my project to use an old set of organ foot pedals into a midi controller.

I have the device working as it should as a single octave device with the octave called in variables section before void setup() using NoteButton.

I want to create an octaveUp and octaveDown button that change the octaves called in NoteButton but it seems like that would have to be in the loop section as it is in several examples of edge detection, debouncing, etcetera.

The below code includes the attempt to add the octave using the variable "count." The code is not complete because it will not compile when I change the number 4 in the NoteButton section to "count."

Should the NoteButton section be moved into the loop or is there a more graceful way to do this?

Thanks in advance.
```cpp
#include <Control_Surface.h> // Include the Control Surface library
 
// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;

using namespace MIDI_Notes;
 
const auto channel = CHANNEL_1;
const uint8_t velocity = 0x7F;

// Instantiate an array of NoteButton objects
NoteButton buttons[] = {
  {23, {note(C, 4), channel}, velocity}, // { pin, {address}, velocity } I would like to be able to change t he number 4 up and down using two bottons
  {25, {note(Db, 4), channel}, velocity},
  {27, {note(D, 4), channel}, velocity},
  {29, {note(Eb, 4), channel}, velocity},
  {31, {note(E, 4), channel}, velocity},
  {33, {note(F, 4), channel}, velocity},
  {35, {note(Gb, 4), channel}, velocity},
  {37, {note(G, 4), channel}, velocity},
  {39, {note(Ab, 4), channel}, velocity},
  {41, {note(A, 4), channel}, velocity},
  {43, {note(Bb, 4), channel}, velocity},
  {45, {note(B, 4), channel}, velocity},
  {47, {note(C, 5), channel}, velocity},


};


// Instantiate a CCPotentiometer object
CCPotentiometer modwheel = {
  A1,                                   // Analog pin connected to potentiometer
  {MIDI_CC::Modulation_Wheel, CHANNEL_1}, // Channel volume of channel 1
};


// Instantiate a PBPotentiometer object
PBPotentiometer pitchbender = {
  A0,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1
  
};

//Added code for the buttons
const byte buttonPins[] = {6, 7};
byte prevButtonStates[] = {HIGH, HIGH};
int increments[] = {1, -1};
int count = 0;
//End of added code for the buttons

 
void setup() {  
  Control_Surface.begin(); // Initialize Control Surface

//Added code for the buttons
{
  Serial.begin(115200);
  for (int button = 0; button < 2; button++)
  {
    pinMode(buttonPins[button], INPUT_PULLUP);
  }
}
//End of added code for the buttons
  
}
 
void loop() {
  Control_Surface.loop(); // Update the Control Surface

{

//Added code for the buttons  
  for (int button = 0; button < 2; button++)
  {
    byte currentButtonState = digitalRead(buttonPins[button]);
    if (currentButtonState != prevButtonStates[button] && currentButtonState == LOW)
    {
      count += increments[button];
      Serial.println(count);
    }
    prevButtonStates[button] = currentButtonState;
  }
}
//End of added code for the buttons
}
```

**Comment by dstoddart** [comment]
One additional note, my goal is something similar to discussion post 414 but using banks seems like a significant departure from the NoteButtons approach.

**Comment by tttapa** [comment]
I think the [Transposer](https://tttapa.github.io/Control-Surface-doc/Doxygen/da/d2d/Transposer_8ino-example.html) example does what you're asking.

To change the notes one octave at a time, use
```cpp
Transposer<-1, +1> transposer(12);
```
(because there are 12 semitones in an octave).

Keep in mind that the addresses you use when initializing the buttons are evaluated and then saved by value. So if you initialize a button using `{23, {note(C, count), channel}, velocity}`, the address will always contain the values `count` and `channel` had at the point of initialization. If you change `count` later in your code, it won't affect the address of the button.

**Reply by dstoddart** [reply]
That's exactly it! I will modify the code and let everyone know how it goes.  Once again, great job on the library!

**Reply by dstoddart** [reply]
Just to confirm, this worked exactly as I hoped. Thanks!

**Comment by matador911** [comment]
/// Get the first MIDI note in the given octave.
constexpr int8_t octave(int8_t numOctave) { return 12 * (numOctave + 1); }

/// Get the MIDI note in the given octave.
constexpr int8_t note(int8_t note, int8_t numOctave) {
    return note + octave(numOctave);
